### ISSUE TITLE
Single file format (again)

### ISSUE THREAD

**Issue by vkbo** [vkbo/novelWriter#977]
The single file format has been discussed before. There are three key benefits of the current structure of saving each text document and meta data unit as a separate file. Especially when it comes to the project's text itself:

1. It is safer against data loss.
2. It is very file sync friendly.
3. It is version control friendly.

**Issue:** The fact that all the data is stored in plain text files seems to imply to many users that the data is open to be edited externally, and therefore the project folder should be easy to navigate for that purpose. But the project folder is intended as a file database, not as a folder with files to be manually accessed. Because of the possibility of manual edits, a lot of extra care has to be taken when reading the data in order to account for user-introduced errors.

A few conditions must apply to a single file format:

* It must be optional (although perhaps default) for the foreseeable future.
* In order to satisfy point 1 above, the file format must retain the individual document read/write properties of the current setup, and not rely on an in-memory data buffer of all project content between open and save/close.
* The file sync stability of point 2 must be retained.
* The version control support can only be retained by using a single flat plain text file format, and even that is not ideal. Point 3 is _not_ a requirement. For users who requires this, opting out of the single file format is a reasonable solution.

**Implementation**

I've considered a few options before, detailed in #259. While using a read/write database as storage would solve the in-place issue of point 1, database files really aren't file sync friendly or even file sync safe, which breaks with point 2.

The one solution that stands out as the simplest solution is to keep the project as an archive file. While an archive is not really writable in the sense demanded by point 1, a temporary workspace can bypass this issue: When the project is opened, it can be extracted into a temp folder, and written back as an archive when the project is saved.

The additional benefit of this implementation is that it avoids the need to make any fundamental changes to how novelWriter reads and writes data during the writing session. All that is needed is to inject an extra step in the open and save process.

Implementation steps:

* The archive I/O itself will be implemented as a [Python ZipFile](https://docs.python.org/3/library/zipfile.html) object.
* Each project will be assigned a UUID when it is first saved as an archive. The UUID will be saved in the "comment" section of the archive, which can contain 64kb of data.
* On open, the archive will be extracted to a folder under the novelWriter app data location (OS dependant). The folder will have the UUID as its name. This means it will be possible to recognise and recover a previous interrupted session left by, for instance, an app crash or a computer crash.
* On save, the content of the UUID folder will be saved to a new zip file, and then it will replace the previous file.
* Optional: Sanitise the file list on read/write to the zip file. This will make it easier to clean out deprecated files, and also ensure that files added in other ways to the archive are cleaned out.

This implementation should be fairly file sync friendly as well. At least as much as a Word or Open Document file, which are built in essentially the same way.

**Comment by HeyMyian** [comment]
I was thinking about this yesterday, nice to see it's coming. Are these archives going to differ from those novelWriter writes as backups? It would be interesting if the backup zips could simply be opened with novelWriter without having to manually unpack them first.

**Comment by vkbo** [comment]
Pretty much, yeah. novelWriter is currently running on the new storage class, so extending this feature is pretty simple.

The relationship between backups and the single file format is actually reversed. I wrote a new zip method for the single file that is now only used for backups, so they will in principle be identical. That said, old backup files may not work as project files due to lacking meta data in the older formats. I may add some converter function to handle this though. It's probably a nice feature.

**Comment by awqk** [comment]
Perhaps I am missing something, but this enhancement looks like it's mostly about preventing users from assuming the files in the project directory can be freely edited.

I believe this is a non-issue:

- users already know there's trouble when you modify a file with two programs at the same time. Try it once, you will avoid doing it again.

- entering the bowels of a program's saved files, you know you're on your own. If you're not a complete fool, you have created a backup.

- supporting file sync and version control, novelWriter has to be resilient vs. file changes anyway. Here, novelWriter is already golden, handling file changes nicely even while a project is opened.

I'm not going into the advantages of saving projects in multiple human-readable files, you already know them. With a single file format, you lose these advantages, and get nothing in return.

Maintenance-wise, you trade dropping part of the code watching out for external edits (which is already implemented and apparently working nicely) vs. supporting two file formats (single file vs. multiple file).

In summary, this enhancement sounds like an anti-feature to me. The only good I see coming from this is #1470, as for the rest, I suggest closing this.

**Comment by vkbo** [comment]
My experience with user feedback suggests otherwise. There are a number of checks in the code to mitigate issues caused by people not leaving the project folder alone.

In any case, single file format is much more intuitive to non-technical users, and the maintenance cost is minimal since everything is abstracted away into the Storage class. I already have the implementation, and it works perfectly fine and isn't particularly complicated. I looked a but under the hood at how Libre Office does it, and copied some parts of that.

My implementation preserves the key advantage of robustness since it uses a temp folder in exactly the same way it now uses the project folder as-is. The only difference is the unzip and zip process during open and close. I avoided the option to cache in memory, which is also possible, just more prone to data loss.

Single file projects also have the benefit of being more portable, and sharable (templates).

I do plan to go ahead with this, although I'm in no rush to complete it. Most of the logic is already in the code though.

**Comment by vkbo** [comment]
I've updated the implementation details, again. Not sure when will be a good time to implement this, but I'm trying to make smaller feature releases, so I'll probably keep pushing this ahead for a while still.
